object/slicer/slicer.go

```diff
@@ -311,32 +278,46 @@ func (x *Slicer) Slice(header object.Object, data io.Reader, attributes ...strin
 	childHeader.ResetPreviousID()
 	childHeader.SetChildren(writtenChildren...)
 
-	_, err = writeInMemObject(x.signer, x.w, childHeader, nil, childMeta)
+	_, err = writeInMemObject(signer, x.w, childHeader, nil, childMeta)
 	if err != nil {
 		return rootID, fmt.Errorf("write linking object: %w", err)
 	}
 
 	return rootID, nil
 }
 
-// InitPayloadStream works similar to Slice but provides PayloadWriter allowing
+// InitPayloadStream works similar to [Slicer.Slice] but provides [PayloadWriter] allowing
 // the caller to write data himself.
-func (x *Slicer) InitPayloadStream(attributes ...string) (*PayloadWriter, error) {
+func (x *Slicer) InitPayloadStream(header object.Object, signer neofscrypto.Signer, opts Options) (*PayloadWriter, error) {
+	containerID, owner, err := headerData(header)
+	if err != nil {
+		return nil, err
+	}
+
+	if x.sessionToken != nil {
+		header.SetSessionToken(x.sessionToken)
+		// session issuer is a container owner.
+		issuer := x.sessionToken.Issuer()
+		owner = issuer
+		header.SetOwnerID(&owner)
+	}
+
 	res := &PayloadWriter{
-		stream:       x.w,
-		signer:       x.signer,
-		container:    x.cnr,
-		owner:        x.owner,
-		currentEpoch: x.opts.currentNeoFSEpoch,
-		sessionToken: x.sessionToken,
-		attributes:   attributes,
-		rootMeta:     newDynamicObjectMetadata(x.opts.withHomoChecksum),
-		childMeta:    newDynamicObjectMetadata(x.opts.withHomoChecksum),
-	}
-
-	res.buf.Grow(int(x.childPayloadSizeLimit()))
+		isHeaderWriteStep: true,
+		headerObject:      header,
+		stream:            x.w,
+		signer:            signer,
+		container:         containerID,
+		owner:             owner,
+		currentEpoch:      opts.currentNeoFSEpoch,
+		sessionToken:      x.sessionToken,
+		rootMeta:          newDynamicObjectMetadata(opts.withHomoChecksum),
+		childMeta:         newDynamicObjectMetadata(opts.withHomoChecksum),
+	}
+
+	res.buf.Grow(int(childPayloadSizeLimit(opts)))
 	res.rootMeta.reset()
-	res.currentWriter = newLimitedWriter(io.MultiWriter(&res.buf, &res.rootMeta), x.childPayloadSizeLimit())
+	res.currentWriter = newLimitedWriter(io.MultiWriter(&res.buf, &res.rootMeta), childPayloadSizeLimit(opts))
```

Comment:
`childPayloadSizeLimit(opts)` is called twice in that func

Reply:
Agree, but the func contains only one `if` and `return`, I decided to leave it as is and don't make a new var